### PR TITLE
C++: Add support for createLSParser to the CWE-611 XXE query.

### PR DESCRIPTION
**Pull request by geoffw0**
Add support for `createLSParser` to the CWE-611 XXE query.

This particular extension is straightforward and we already have test cases.

### CODE REVIEW

**Comment by geoffw0** [comment]
No change note required - covered by the note in recent PR https://github.com/github/codeql/pull/8736.

File: cpp/ql/src/Security/CWE/CWE-611/XXE.ql [hunk 1]
```diff
@@ -163,6 +163,13 @@ class ParseFunction extends Function {
   ParseFunction() { this.getClassAndName("parse") instanceof AbstractDOMParserClass }
 }
 
+/**
+ * The `createLSParser` function that returns a newly created `LSParser` object.
+ */
+class CreateLSParser extends Function {
+  CreateLSParser() { this.hasName("createLSParser") }
```
Comment (MathiasVP) [inline cpp/ql/src/Security/CWE/CWE-611/XXE.ql:170]: Is this name too general to just be captured by the function name? If I were to create a function that constructed a class that captured, say, the output of `ls` from the command line I'd call this function `createLSParser` (hm, but maybe with a different casing, though). Should we check either that the return type is of the right type, or that the function lives in some namespace?

**Review by MathiasVP — APPROVED** [review]
LGTM other than a tiny comment.

File: cpp/ql/src/Security/CWE/CWE-611/XXE.ql (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/Security/CWE/CWE-611/XXE.ql:170]: Yeah, I should probably figure out the namespace this should be in and check that as well.

Comment (geoffw0) [reply cpp/ql/src/Security/CWE/CWE-611/XXE.ql]: It turns out the namespace name varies depending on the Xerces version, and its a class method but I'm not entirely confident the class name is stable.  So I've added matching on the return type, which the docs are clear is `DOMLSParser *`.  That should rule out any false matches.

File: cpp/ql/src/Security/CWE/CWE-611/XXE.ql [hunk 2]
```diff
@@ -164,7 +164,17 @@ class ParseFunction extends Function {
 }
 
 /**
- * Configuration for tracking XML objects and their states.
+ * The `createLSParser` function that returns a newly created `LSParser` object.
+ */
+class CreateLSParser extends Function {
+  CreateLSParser() {
+    this.hasName("createLSParser") and
+    this.getType().(PointerType).getBaseType().getName() = "DOMLSParser" // returns a `DOMLSParser *`.
```
Comment (MathiasVP) [inline cpp/ql/src/Security/CWE/CWE-611/XXE.ql:172]: This should make it handle typedefs as well.
```suggestion
    this.getUnderlyingType().(PointerType).getBaseType().getName() = "DOMLSParser" // returns a `DOMLSParser *`.
```

File: cpp/ql/src/Security/CWE/CWE-611/XXE.ql (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-611/XXE.ql]: Fantastic!

File: cpp/ql/src/Security/CWE/CWE-611/XXE.ql (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-611/XXE.ql]: Oops. I think I gave you bad advice. The predicate is called `getUnspecifiedType` (which explains the consistency error reported by ql-for-ql).

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**